### PR TITLE
Add a decorator for reproducing failures with Hypothesis's internal format

### PR DESCRIPTION
**Pull request by DRMacIver**
Seed printing (#860) was added to make it easier to reproduce failures from CI locally. Unfortunately, it doesn't work if you're using the example database on your CI, which is rather counterproductive!

This PR adds a `@reproduce_failure` decorator you can add *temporarily* to your code to reproduce an example from CI or wherever.  This uses the Hypothesis internal database format, encoded in a way that makes it easy to use in your source.

People are discouraged from depending on this format and using it as a substitute for `@example` or the example database by the simple expedient of making a test with this decorator *always* fail. It either fails with the original error, or with a new `DidNotReproduce` error that indicates that the failure has gone away.

Full disclosure: I'm not super happy about this feature, but it fills an important need and I have not come up with any other solution that I dislike less. Suggestions for improving it are very welcome.

### CODE REVIEW

**Comment by DRMacIver** [comment]
Note: The coverage failures are the same as in 1000. I think this must be the result of a recent coverage bug fix, though it's weird we've not seen it before now. Propose solving this problem by merging #1000 first and not worrying about it.

File: src/hypothesis/core.py [hunk 1]
```diff
@@ -812,6 +887,36 @@ def wrapped_test(*arguments, **kwargs):
...
+                except StopTest:
+                    raise DidNotReproduce(
+                        'The shape of the test has changed in some way '
+                        'from where this blob was defined. Are you sure '
+                        "you're running the same test? This could also be "
+                        'caused by using different versions of Hypothesis.'
+                    )
+                except UnsatisfiedAssumption:
+                    raise DidNotReproduce(
+                        'The test data failed to satisfy an assumption in the '
```
Comment (Zac-HD) [inline src/hypothesis/core.py]: I'm tempted to include a hash of `inspect.getsource`, so we can reject known-different tests.

For usablilty, skip the local check if CI can't find the source (write all-zero 'hash'), and if local can't find source assume it's the same.

**Comment by DRMacIver** [comment]
> General theme of comments: we should encode more information in the blob, so we can detect things that would be expected to break reproducibility

I thought about this, but the problem is that the more data we encode in the blob, the noisier the entire experience gets!

One thing I was vaguely thinking of is that we do have 7 bits left in the leading byte though. We *could* just take all of these things we want to deduplicate on and hash them together and take the low 7 bits. It doesn't give us perfect detection, but it at least makes mistakes less likely.

File: docs/reproducing.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  from unittest import TestCase
+
+  class TestThings(TestCase):
+      @given(text())
+      @example("Hello world")
+      @example(x="Some very long string")
+      def test_some_code(self, x):
+          assert True
+
+It is *not* permitted for a single example to be a mix of positional and
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply docs/reproducing.rst]: True. I think when that was originally written it *wasn't* the same for `@given` and I just CBA to make the logic work in both cases. Will update.

**Comment by Zac-HD** [comment]
From my own debugging experience, highly specific reporting is *really* valuable - I'd be delighted to get version checks and a quick sanity-check on the test function, especially at the low cost of copy-pasting a few more characters.  In short, debugging already sucks when you're this far, so we should make the worst-case as good as possible even if that's not ideal in more common cases.

The idea here is that the two most common reasons for this to fail are that (1) the environments are different, or (2) the string has been corrupted.  In the latter case, recording the length (mod 256) and parity (ie sum mod 256) in two bytes / three b64 characters seems like a bargain to me - I'm just not so sensitive to length that that's ever a problem, and almost-certainly detecting malformed input would be very reassuring.

Spending a byte on Python version (27, 34, etc), two on Hypothesis (minor/patch), and up to four on a source hash also seems very useful to me - that's still only 11 characters as base64, and avoids several frustrations entirely.

**Comment by DRMacIver** [comment]
OK. I think the tests should be passing now - there turned out to be a fun (sigh) coverage bug which I had to work around. I've made a couple significant changes:

* We now run in a relatively pure randomized mode as long as we're checking health checks. This means that we now can run health checks reliably with `@seed`, so we do print seeds in that case.
* Hypothesis version is included in the call to reproduce_failure (explicitly, not encoded!). We could add more information, but I think this is a decent first pass. We're also more explicit about the fact that the reproduce_failure API is in no way ever going to be backwards compatible because it's temporary (so it's easy to add more arguments).

I'm still not 100% happy with this solution but:

1. I still don't have a better one.
2. It's documented as completely incompatible between Hypothesis versions, so it's no great hardship to kill it if we come up with one.

**Comment by DRMacIver** [comment]
> . Could these coverage problems might be related to #997?

No, all that code is disabled when running under coverage.

File: tests/cover/test_seed_printing.py [hunk 3]
```diff
@@ -23,15 +23,18 @@
...
+from tests.common.utils import all_values, capture_out
+from hypothesis.database import InMemoryExampleDatabase
 from hypothesis.internal.compat import hrange
 
 
 @pytest.mark.parametrize('in_pytest', [False, True])
 @pytest.mark.parametrize('fail_healthcheck', [False, True])
-def test_prints_seed_on_exception(monkeypatch, in_pytest, fail_healthcheck):
+def test_prints_seed_only_on_healthcheck(
+    monkeypatch, in_pytest, fail_healthcheck
```
[the remark this answers is not in the document]
Comment (DRMacIver) [reply tests/cover/test_seed_printing.py]: The test name got longer to match the new behaviour, so it now exceeds the 80 character limit if the arguments are not on a new line.

[the remark this answers is not in the document]
Comment (Zac-HD) [reply tests/cover/test_seed_printing.py]: Ugh. that's what I get for writing late-night reviews .

**Comment by DRMacIver** [comment]
> (I still don't find it aesthetic, but I think that's just me now)

No, I'm definitely not wild about the aesthetics either. This isn't my favourite solution to the problem so much as the least bad one I've been able to think of! I'd rather have it solved in a way that is slightly ugly than not have it solved though, and there's a certain amount of disposability built into this feature if we come up with a better solution.

**Pull request merged**